File: benchmarks/JavaScript/havlak.js
// @ts-check
// Adapted based on SOM and Java benchmark.
//  Copyright 2011 Google Inc.
//
//      Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//      You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//      See the License for the specific language governing permissions and
//          limitations under the License.

const { Benchmark } = require('./benchmark');
const {
  Vector, Set, IdentitySet, IdentityDictionary
} = require('./som');

class BasicBlock {
  constructor(name) {
    this.name = name;
    this.inEdges = new Vector(2);
    this.outEdges = new Vector(2);
  }

  getInEdges() {
    return this.inEdges;
  }

  getOutEdges() {
    return this.outEdges;
  }

  getNumPred() {
    return this.inEdges.size();
  }

  addOutEdge(to) {
    return this.outEdges.append(to);
  }

  addInEdge(from) {
    return this.inEdges.append(from);
  }

  customHash() {
    return this.name;
  }
}

function BasicBlockEdge(cfg, fromName, toName) {
  this.from = cfg.createNode(fromName);
  this.to = cfg.createNode(toName);

  this.from.addOutEdge(this.to);
  this.to.addInEdge(this.from);

  cfg.addEdge(this);
}

class ControlFlowGraph {
  constructor() {
    this.startNode = null;
    this.basicBlockMap = new Vector();
    this.edgeList = new Vector();
  }

  createNode(name) {
    let node;
    if (this.basicBlockMap.at(name)) {
      node = this.basicBlockMap.at(name);
    } else {
      node = new BasicBlock(name);
      this.basicBlockMap.atPut(name, node);
    }

    if (this.getNumNodes() === 1) {
      this.startNode = node;
    }
    return node;
  }

  addEdge(edge) {
    this.edgeList.append(edge);
  }

  getNumNodes() {
    return this.basicBlockMap.size();
  }

  getStartBasicBlock() {
    return this.startNode;
  }

  getBasicBlocks() {
    return this.basicBlockMap;
  }
}

class SimpleLoop {
  constructor(bb, isReducible) {
    this.isReducible = isReducible;
    this.parent = null;
    this.isRoot_ = false;
    this.nestingLevel = 0;
    this.depthLevel = 0;
    this.basicBlocks = new IdentitySet();
    this.children = new IdentitySet();

    if (bb) {
      this.basicBlocks.add(bb);
    }
    this.header = bb;
  }

  addNode(bb) {
    this.basicBlocks.add(bb);
  }

  addChildLoop(loop) {
    this.children.add(loop);
  }

  // Getters/Setters
  getChildren() {
    return this.children;
  }

  getParent() {
    return this.parent;
  }

  getNestingLevel() {
    return this.nestingLevel;
  }

  isRoot() {
    return this.isRoot_;
  }

  setParent(parent) {
    this.parent = parent;
    this.parent.addChildLoop(this);
  }

  setIsRoot() {
    this.isRoot_ = true;
  }

  setCounter(value) {
    this.counter = value;
  }

  setNestingLevel(level) {
    this.nestingLevel = level;
    if (level === 0) {
      this.setIsRoot();
    }
  }

  setDepthLevel(level) {
    this.depthLevel = level;
  }
}

class LoopStructureGraph {
  constructor() {
    this.loopCounter = 0;
    this.loops = new Vector();
    this.root = new SimpleLoop(null, true);
    this.root.setNestingLevel(0);
    this.root.setCounter(this.loopCounter);
    this.loopCounter += 1;
    this.loops.append(this.root);
  }

  createNewLoop(bb, isReducible) {
    const loop = new SimpleLoop(bb, isReducible);
    loop.setCounter(this.loopCounter);
    this.loopCounter += 1;
    this.loops.append(loop);
    return loop;
  }

  calculateNestingLevel() {
    // link up all 1st level loops to artificial root node.
    this.loops.forEach((liter) => {
      if (!liter.isRoot()) {
        if (!liter.getParent()) {
          liter.setParent(this.root);
        }
      }
    });

    // recursively traverse the tree and assign levels.
    this.calculateNestingLevelRec(this.root, 0);
  }

  calculateNestingLevelRec(loop, depth) {
    loop.setDepthLevel(depth);
    loop.getChildren().forEach((liter) => {
      this.calculateNestingLevelRec(liter, depth + 1);

      loop.setNestingLevel(
        Math.max(
          loop.getNestingLevel(),
          1 + liter.getNestingLevel()
        )
      );
    });
  }

  getNumLoops() {
    return this.loops.size();
  }
}

class UnionFindNode {
  // Initialize this node.
  initNode(bb, dfsNumber) {
    this.parent = this;
    this.bb = bb;
    this.dfsNumber = dfsNumber;
    this.loop = null;
  }

  // Union/Find Algorithm - The find routine.
  //
  // Implemented with Path Compression (inner loops are only
  // visited and collapsed once, however, deep nests would still
  // result in significant traversals).
  //
  findSet() {
    const nodeList = new Vector();
    let node = this;

    while (node !== node.parent) {
      if (node.parent !== node.parent.parent) {
        nodeList.append(node);
      }
      node = node.parent;
    }

    // Path Compression, all nodes' parents point to the 1st level parent.
    nodeList.forEach((iter) => iter.union(this.parent));
    return node;
  }

  // Union/Find Algorithm - The union routine.
  //
  // Trivial. Assigning parent pointer is enough,
  // we rely on path compression.
  //
  union(basicBlock) {
    this.parent = basicBlock;
  }

  // Getters/Setters
  //
  getBb() {
    return this.bb;
  }

  getLoop() {
    return this.loop;
  }

  getDfsNumber() {
    return this.dfsNumber;
  }

  setLoop(loop) {
    this.loop = loop;
  }
}

const UNVISITED = 2147483647; // Marker for uninitialized nodes.
const MAXNONBACKPREDS = (32 * 1024); // Safeguard against pathological algorithm behavior.

class HavlakLoopFinder {
  constructor(cfg, lsg) {
    this.nonBackPreds = new Vector();
    this.backPreds = new Vector();
    this.number = new IdentityDictionary();
    this.maxSize = 0;

    this.header = null;
    this.type = null;
    this.last = null;
    this.nodes = null;

    this.cfg = cfg;
    this.lsg = lsg;
  }

  // As described in the paper, determine whether a node 'w' is a
  // "true" ancestor for node 'v'.
  //
  // Dominance can be tested quickly using a pre-order trick
  // for depth-first spanning trees. This is why DFS is the first
  // thing we run below.
  isAncestor(w, v) {
    return w <= v && v <= this.last[w];
  }

  // DFS - Depth-First-Search
  //
  // DESCRIPTION:
  // Simple depth first traversal along out edges with node numbering.
  doDFS(currentNode, current) {
    this.nodes[current].initNode(currentNode, current);
    this.number.atPut(currentNode, current);

    let lastId = current;
    const outerBlocks = currentNode.getOutEdges();

    for (let i = 0; i < outerBlocks.size(); i += 1) {
      const target = outerBlocks.at(i);
      if (this.number.at(target) === UNVISITED) {
        lastId = this.doDFS(target, lastId + 1);
      }
    }

    this.last[current] = lastId;
    return lastId;
  }

  initAllNodes() {
    // Step a:
    //   - initialize all nodes as unvisited.
    //   - depth-first traversal and numbering.
    //   - unreached BB's are marked as dead.
    this.cfg.getBasicBlocks().forEach(
      (bb) => { this.number.atPut(bb, UNVISITED); }
    );

    this.doDFS(this.cfg.getStartBasicBlock(), 0);
  }

  identifyEdges(size) {
    // Step b:
    //   - iterate over all nodes.
    //
    //   A backedge comes from a descendant in the DFS tree, and non-backedges
    //   from non-descendants (following Tarjan).
    //
    //   - check incoming edges 'v' and add them to either
    //     - the list of backedges (backPreds) or
    //     - the list of non-backedges (nonBackPreds)
    for (let w = 0; w < size; w += 1) {
      this.header[w] = 0;
      this.type[w] = 'BB_NONHEADER';

      const nodeW = this.nodes[w].getBb();
      if (!nodeW) {
        this.type[w] = 'BB_DEAD';
      } else {
        this.processEdges(nodeW, w);
      }
    }
  }

  processEdges(nodeW, w) {
    if (nodeW.getNumPred() > 0) {
      nodeW.getInEdges().forEach((nodeV) => {
        const v = this.number.at(nodeV);
        if (v !== UNVISITED) {
          if (this.isAncestor(w, v)) {
            this.backPreds.at(w).append(v);
          } else {
            this.nonBackPreds.at(w).add(v);
          }
        }
      });
    }
  }

  // Find loops and build loop forest using Havlak's algorithm, which
  // is derived from Tarjan. Variable names and step numbering has
  // been chosen to be identical to the nomenclature in Havlak's
  // paper (which, in turn, is similar to the one used by Tarjan).
  findLoops() {
    if (!this.cfg.getStartBasicBlock()) {
      return;
    }

    const size = this.cfg.getNumNodes();

    this.nonBackPreds.removeAll();
    this.backPreds.removeAll();
    this.number.removeAll();
    if (size > this.maxSize) {
      this.header = new Array(size);
      this.type = new Array(size);
      this.last = new Array(size);
      this.nodes = new Array(size);
      this.maxSize = size;
    }

    for (let i = 0; i < size; i += 1) {
      this.nonBackPreds.append(new Set());
      this.backPreds.append(new Vector());
      this.nodes[i] = new UnionFindNode();
    }

    this.initAllNodes();
    this.identifyEdges(size);

    // Start node is root of all other loops.
    this.header[0] = 0;

    // Step c:
    //
    // The outer loop, unchanged from Tarjan. It does nothing except
    // for those nodes which are the destinations of backedges.
    // For a header node w, we chase backward from the sources of the
    // backedges adding nodes to the set P, representing the body of
    // the loop headed by w.
    //
    // By running through the nodes in reverse of the DFST preorder,
    // we ensure that inner loop headers will be processed before the
    // headers for surrounding loops.
    //
    for (let w = size - 1; w >= 0; w -= 1) {
      // this is 'P' in Havlak's paper
      const nodePool = new Vector();

      const nodeW = this.nodes[w].getBb();
      if (nodeW) {
        this.stepD(w, nodePool);

        // Copy nodePool to workList.
        const workList = new Vector();
        nodePool.forEach((niter) => workList.append(niter));

        if (nodePool.size() !== 0) {
          this.type[w] = 'BB_REDUCIBLE';
        }

        // work the list...
        while (!workList.isEmpty()) {
          const x = workList.removeFirst();

          // Step e:
          //
          // Step e represents the main difference from Tarjan's method.
          // Chasing upwards from the sources of a node w's backedges. If
          // there is a node y' that is not a descendant of w, w is marked
          // the header of an irreducible loop, there is another entry
          // into this loop that avoids w.

          // The algorithm has degenerated. Break and
          // return in this case.
          const nonBackSize = this.nonBackPreds.at(x.getDfsNumber()).size();
          if (nonBackSize > MAXNONBACKPREDS) {
            return;
          }
          this.stepEProcessNonBackPreds(w, nodePool, workList, x);
        }

        // Collapse/Unionize nodes in a SCC to a single node
        // For every SCC found, create a loop descriptor and link it in.
        //
        if ((nodePool.size() > 0) || (this.type[w] === 'BB_SELF')) {
          const loop = this.lsg.createNewLoop(nodeW, this.type[w] !== 'BB_IRREDUCIBLE');
          this.setLoopAttributes(w, nodePool, loop);
        }
      }
    } // Step c
  } // findLoops

  stepEProcessNonBackPreds(w, nodePool, workList, x) {
    this.nonBackPreds.at(x.getDfsNumber()).forEach((iter) => {
      const y = this.nodes[iter];
      const ydash = y.findSet();

      if (!this.isAncestor(w, ydash.getDfsNumber())) {
        this.type[w] = 'BB_IRREDUCIBLE';
        this.nonBackPreds.at(w).add(ydash.getDfsNumber());
      } else if (ydash.getDfsNumber() !== w) {
        if (!nodePool.hasSome((e) => e === ydash)) {
          workList.append(ydash);
          nodePool.append(ydash);
        }
      }
    });
  }

  setLoopAttributes(w, nodePool, loop) {
    // At this point, one can set attributes to the loop, such as:
    //
    // the bottom node:
    //    iter  = backPreds[w].begin();
    //    loop bottom is: nodes[iter].node);
    //
    // the number of backedges:
    //    backPreds[w].size()
    //
    // whether this loop is reducible:
    //    type[w] != BasicBlockClass.BB_IRREDUCIBLE
    //
    this.nodes[w].setLoop(loop);

    nodePool.forEach((node) => {
      // Add nodes to loop descriptor.
      this.header[node.getDfsNumber()] = w;
      node.union(this.nodes[w]);

      // Nested loops are not added, but linked together.
      if (node.getLoop()) {
        node.getLoop().setParent(loop);
      } else {
        loop.addNode(node.getBb());
      }
    });
  }

  stepD(w, nodePool) {
    this.backPreds.at(w).forEach((v) => {
      if (v !== w) {
        nodePool.append(this.nodes[v].findSet());
      } else {
        this.type[w] = 'BB_SELF';
      }
    });
  }
}

class LoopTesterApp {
  constructor() {
    this.cfg = new ControlFlowGraph();
    this.lsg = new LoopStructureGraph();
    this.cfg.createNode(0);
  }

  // Create 4 basic blocks, corresponding to and if/then/else clause
  // with a CFG that looks like a diamond
  buildDiamond(start) {
    const bb0 = start;
    new BasicBlockEdge(this.cfg, bb0, bb0 + 1);
    new BasicBlockEdge(this.cfg, bb0, bb0 + 2);
    new BasicBlockEdge(this.cfg, bb0 + 1, bb0 + 3);
    new BasicBlockEdge(this.cfg, bb0 + 2, bb0 + 3);

    return bb0 + 3;
  }

  // Connect two existing nodes
  buildConnect(start, end) {
    new BasicBlockEdge(this.cfg, start, end);
  }

  // Form a straight connected sequence of n basic blocks
  buildStraight(start, n) {
    for (let i = 0; i < n; i += 1) {
      this.buildConnect(start + i, start + i + 1);
    }
    return start + n;
  }

  // Construct a simple loop with two diamonds in it
  buildBaseLoop(from) {
    const header = this.buildStraight(from, 1);
    const diamond1 = this.buildDiamond(header);
    const d11 = this.buildStraight(diamond1, 1);
    const diamond2 = this.buildDiamond(d11);

    let footer = this.buildStraight(diamond2, 1);
    this.buildConnect(diamond2, d11);
    this.buildConnect(diamond1, header);

    this.buildConnect(footer, from);
    footer = this.buildStraight(footer, 1);
    return footer;
  }

  main(numDummyLoops, findLoopIterations, parLoops, pparLoops, ppparLoops) {
    this.constructSimpleCFG();
    this.addDummyLoops(numDummyLoops);
    this.constructCFG(parLoops, pparLoops, ppparLoops);

    // Performing Loop Recognition, 1 Iteration, then findLoopIteration
    this.findLoops(this.lsg);
    for (let i = 0; i < findLoopIterations; i += 1) {
      this.findLoops(new LoopStructureGraph());
    }

    this.lsg.calculateNestingLevel();
    return [this.lsg.getNumLoops(), this.cfg.getNumNodes()];
  }

  constructCFG(parLoops, pparLoops, ppparLoops) {
    let n = 2;

    for (let parlooptrees = 0; parlooptrees < parLoops; parlooptrees += 1) {
      this.cfg.createNode(n + 1);
      this.buildConnect(2, n + 1);
      n += 1;

      for (let i = 0; i < pparLoops; i += 1) {
        const top = n;
        n = this.buildStraight(n, 1);
        for (let j = 0; j < ppparLoops; j += 1) {
          n = this.buildBaseLoop(n);
        }
        const bottom = this.buildStraight(n, 1);
        this.buildConnect(n, top);
        n = bottom;
      }
      this.buildConnect(n, 1);
    }
  }

  addDummyLoops(numDummyLoops) {
    for (let dummyloop = 0; dummyloop < numDummyLoops; dummyloop += 1) {
      this.findLoops(this.lsg);
    }
  }

  findLoops(loopStructure) {
    const finder = new HavlakLoopFinder(this.cfg, loopStructure);
    finder.findLoops();
  }

  constructSimpleCFG() {
    this.cfg.createNode(0);
    this.buildBaseLoop(0);
    this.cfg.createNode(1);
    new BasicBlockEdge(this.cfg, 0, 2);
  }
}

class Havlak extends Benchmark {
  innerBenchmarkLoop(innerIterations) {
    return this.verifyResult(
      (new LoopTesterApp()).main(
        innerIterations,
        50,
        10 /* was 100 */,
        10,
        5
      ),
      innerIterations
    );
  }

  verifyResult(result, innerIterations) {
    if (innerIterations === 15000) { return result[0] === 46602 && result[1] === 5213; }
    if (innerIterations === 1500) { return result[0] === 6102 && result[1] === 5213; }
    if (innerIterations === 150) { return result[0] === 2052 && result[1] === 5213; }
    if (innerIterations === 15) { return result[0] === 1647 && result[1] === 5213; }
    if (innerIterations === 1) { return result[0] === 1605 && result[1] === 5213; }

    process.stdout.write(`No verification result for ${innerIterations} found`);
    process.stdout.write(`Result is: ${result[0]}, ${result[1]}`);
    return false;
  }
}

exports.newInstance = () => new Havlak();

File: benchmarks/Python/havlak.py
# Adapted based on SOM benchmark.
# Copyright 2011 Google Inc.
#
#     Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
#     You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
#     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#     See the License for the specific language governing permissions and
#         limitations under the License.
from enum import Enum

import sys

from benchmark import Benchmark
from som.identity_dictionary import IdentityDictionary
from som.identity_set import IdentitySet
from som.set import Set
from som.vector import Vector

# Havlak needs more stack space in CPython
sys.setrecursionlimit(1500)


class Havlak(Benchmark):
    def inner_benchmark_loop(self, inner_iterations):
        return self._verify_result(
            _LoopTesterApp().main(inner_iterations, 50, 10, 10, 5), inner_iterations
        )

    @staticmethod
    def _verify_result(result, inner_iterations):
        if inner_iterations == 15_000:
            return result[0] == 46_602 and result[1] == 5213
        if inner_iterations == 1_500:
            return result[0] == 6_102 and result[1] == 5213
        if inner_iterations == 150:
            return result[0] == 2_052 and result[1] == 5213
        if inner_iterations == 15:
            return result[0] == 1_647 and result[1] == 5213
        if inner_iterations == 1:
            return result[0] == 1_605 and result[1] == 5213

        print("No verification result for " + str(inner_iterations) + " found")
        print("Result is: " + str(result[0]) + ", " + str(result[1]))
        return False

    def benchmark(self):
        raise Exception("should not be reached")

    def verify_result(self, result):
        raise Exception("should not be reached")


class _BasicBlock:
    def __init__(self, name):
        self._name = name
        self.in_edges = Vector(2)
        self.out_edges = Vector(2)

    def get_num_pred(self):
        return self.in_edges.size()

    def add_out_edge(self, to):
        self.out_edges.append(to)

    def add_in_edge(self, from_):
        self.in_edges.append(from_)

    def custom_hash(self):
        return self._name


class _BasicBlockEdge:
    def __init__(self, cfg, from_name, to_name):
        self._from = cfg.create_node(from_name)
        self._to = cfg.create_node(to_name)

        self._from.add_out_edge(self._to)
        self._to.add_in_edge(self._from)

        cfg.add_edge(self)


class _ControlFlowGraph:
    def __init__(self):
        self.start_basic_block = None
        self.basic_blocks = Vector()
        self._edge_list = Vector()

    def create_node(self, name):
        if self.basic_blocks.at(name):
            node = self.basic_blocks.at(name)
        else:
            node = _BasicBlock(name)
            self.basic_blocks.at_put(name, node)

        if self.num_nodes() == 1:
            self.start_basic_block = node
        return node

    def add_edge(self, edge):
        self._edge_list.append(edge)

    def num_nodes(self):
        return self.basic_blocks.size()


class _LoopStructureGraph:
    def __init__(self):
        self._loop_counter = 0
        self._loops = Vector()
        self._root = _SimpleLoop(None, True)
        self._root.set_nesting_level(0)
        self._root.counter = self._loop_counter
        self._loop_counter += 1
        self._loops.append(self._root)

    def create_new_loop(self, bb, is_reducible):
        loop = _SimpleLoop(bb, is_reducible)
        loop.counter = self._loop_counter
        self._loop_counter += 1
        self._loops.append(loop)
        return loop

    def calculate_nesting_level(self):
        def each(liter):
            if not liter.is_root:
                if liter.parent is None:
                    liter.set_parent(self._root)

        self._loops.for_each(each)
        self._calculate_nesting_level_rec(self._root, 0)

    def _calculate_nesting_level_rec(self, loop, depth):
        loop.depth_level = depth

        def each(liter):
            self._calculate_nesting_level_rec(liter, depth + 1)
            loop.set_nesting_level(max(loop.nesting_level, 1 + liter.nesting_level))

        loop.children.for_each(each)

    def num_loops(self):
        return self._loops.size()


class _SimpleLoop:
    def __init__(self, bb, is_reducible):
        self._is_reducible = is_reducible
        self.parent = None
        self.is_root = False
        self.nesting_level = 0
        self._depth_level = 0
        self._counter = 0
        self._basic_blocks = IdentitySet()
        self.children = IdentitySet()

        if bb is not None:
            self._basic_blocks.add(bb)

        self._header = bb

    def add_node(self, bb):
        self._basic_blocks.add(bb)

    def add_child_loop(self, loop):
        self.children.add(loop)

    def set_parent(self, parent):
        self.parent = parent
        self.parent.add_child_loop(self)

    def set_nesting_level(self, level):
        self.nesting_level = level
        if level == 0:
            self.is_root = True


class _UnionFindNode:
    def __init__(self):
        self.parent = None
        self.bb = None
        self.dfs_number = 0
        self.loop = None

    def init_node(self, bb, dfs_number):
        self.parent = self
        self.bb = bb
        self.dfs_number = dfs_number
        self.loop = None

    def find_set(self):
        node_list = Vector()

        node = self
        while node is not node.parent:
            if node.parent is not node.parent.parent:
                node_list.append(node)
            node = node.parent

        node_list.for_each(lambda i: i.union(self.parent))
        return node

    def union(self, basic_block):
        self.parent = basic_block


class _LoopTesterApp:
    def __init__(self):
        self._cfg = _ControlFlowGraph()
        self._lsg = _LoopStructureGraph()
        self._cfg.create_node(0)

    def _build_diamond(self, start):
        bb0 = start
        _BasicBlockEdge(self._cfg, bb0, bb0 + 1)
        _BasicBlockEdge(self._cfg, bb0, bb0 + 2)
        _BasicBlockEdge(self._cfg, bb0 + 1, bb0 + 3)
        _BasicBlockEdge(self._cfg, bb0 + 2, bb0 + 3)
        return bb0 + 3

    def _build_connect(self, start, end_):
        _BasicBlockEdge(self._cfg, start, end_)

    def _build_straight(self, start, n):
        for i in range(n):
            self._build_connect(start + i, start + i + 1)
        return start + n

    def _build_base_loop(self, from_):
        header = self._build_straight(from_, 1)
        diamond1 = self._build_diamond(header)
        d11 = self._build_straight(diamond1, 1)
        diamond2 = self._build_diamond(d11)
        footer = self._build_straight(diamond2, 1)
        self._build_connect(diamond2, d11)
        self._build_connect(diamond1, header)

        self._build_connect(footer, from_)
        footer = self._build_straight(footer, 1)
        return footer

    def main(
        self, num_dummy_loops, find_loop_iterations, par_loops, ppar_loops, pppar_loops
    ):
        self._construct_simple_cfg()
        self._add_dummy_loops(num_dummy_loops)
        self._construct_cfg(par_loops, ppar_loops, pppar_loops)

        self._find_loops(self._lsg)
        for _ in range(find_loop_iterations):
            self._find_loops(_LoopStructureGraph())

        self._lsg.calculate_nesting_level()
        return [self._lsg.num_loops(), self._cfg.num_nodes()]

    def _construct_cfg(self, par_loops, ppar_loops, pppar_loops):
        n = 2

        for _ in range(par_loops):
            self._cfg.create_node(n + 1)
            self._build_connect(2, n + 1)
            n += 1

            for _ in range(ppar_loops):
                top = n
                n = self._build_straight(n, 1)
                for _ in range(pppar_loops):
                    n = self._build_base_loop(n)
                bottom = self._build_straight(n, 1)
                self._build_connect(n, top)
                n = bottom

            self._build_connect(n, 1)

    def _add_dummy_loops(self, num_dummy_loops):
        for _ in range(num_dummy_loops):
            self._find_loops(self._lsg)

    def _find_loops(self, loop_structure):
        finder = _HavlakLoopFinder(self._cfg, loop_structure)
        finder.find_loops()

    def _construct_simple_cfg(self):
        self._cfg.create_node(0)
        self._build_base_loop(0)
        self._cfg.create_node(1)
        _BasicBlockEdge(self._cfg, 0, 2)


_UNVISITED = 2_147_483_647
_MAXNONBACKPREDS = 32 * 1024


class _BasicBlockClass(Enum):
    BB_TOP = 0  # uninitialized
    BB_NONHEADER = 1  # a regular BB
    BB_REDUCIBLE = 2  # reducible loop
    BB_SELF = 3  # single BB loop
    BB_IRREDUCIBLE = 4  # irreducible loop
    BB_DEAD = 5  # a dead BB
    BB_LAST = 6  # Sentinel


class _HavlakLoopFinder:
    def __init__(self, cfg, lsg):
        self._cfg = cfg
        self._lsg = lsg
        self._non_back_preds = Vector()
        self._back_preds = Vector()
        self._number = IdentityDictionary()
        self._max_size = 0
        self._header = None
        self._type = None
        self._last = None
        self._nodes = None

    def _is_ancestor(self, w, v):
        return w <= v and v <= self._last[w]

    def _do_dfs(self, current_node, current):
        self._nodes[current].init_node(current_node, current)
        self._number.at_put(current_node, current)

        last_id = current
        outer_blocks = current_node.out_edges

        def each(target):
            nonlocal last_id
            if self._number.at(target) == _UNVISITED:
                last_id = self._do_dfs(target, last_id + 1)

        outer_blocks.for_each(each)

        self._last[current] = last_id
        return last_id

    def _init_all_nodes(self):
        self._cfg.basic_blocks.for_each(lambda bb: self._number.at_put(bb, _UNVISITED))

        self._do_dfs(self._cfg.start_basic_block, 0)

    def _identify_edges(self, size):
        for w in range(size):
            self._header[w] = 0
            self._type[w] = _BasicBlockClass.BB_NONHEADER

            node_w = self._nodes[w].bb
            if node_w is None:
                self._type[w] = _BasicBlockClass.BB_DEAD
            else:
                self._process_edges(node_w, w)

    def _process_edges(self, node_w, w):
        if node_w.get_num_pred() > 0:

            def each(node_v):
                v = self._number.at(node_v)
                if v != _UNVISITED:
                    if self._is_ancestor(w, v):
                        self._back_preds.at(w).append(v)
                    else:
                        self._non_back_preds.at(w).add(v)

            node_w.in_edges.for_each(each)

    def find_loops(self):
        if self._cfg.start_basic_block is None:
            return

        size = self._cfg.num_nodes()
        self._non_back_preds.remove_all()
        self._back_preds.remove_all()
        self._number.remove_all()

        if size > self._max_size:
            self._header = [0] * size
            self._type = [None] * size
            self._last = [0] * size
            self._nodes = [None] * size
            self._max_size = size

        for i in range(size):
            self._non_back_preds.append(Set())
            self._back_preds.append(Vector())
            self._nodes[i] = _UnionFindNode()

        self._init_all_nodes()
        self._identify_edges(size)

        self._header[0] = 0

        for w in range(size - 1, -1, -1):
            node_pool = Vector()
            node_w = self._nodes[w].bb
            if node_w is not None:
                self._step_d(w, node_pool)

                work_list = Vector()
                node_pool.for_each(work_list.append)

                if node_pool.size() != 0:
                    self._type[w] = _BasicBlockClass.BB_REDUCIBLE

                while not work_list.is_empty():
                    x = work_list.remove_first()

                    non_back_size = self._non_back_preds.at(x.dfs_number).size()
                    if non_back_size > _MAXNONBACKPREDS:
                        return

                    self._step_e_process_non_back_preds(w, node_pool, work_list, x)

            if node_pool.size() > 0 or self._type[w] == _BasicBlockClass.BB_SELF:
                loop = self._lsg.create_new_loop(
                    node_w, self._type[w] != _BasicBlockClass.BB_IRREDUCIBLE
                )
                self._set_loop_attributes(w, node_pool, loop)

    def _step_e_process_non_back_preds(self, w, node_pool, work_list, x):
        def each(i):
            y = self._nodes[i]
            ydash = y.find_set()

            if not self._is_ancestor(w, ydash.dfs_number):
                self._type[w] = _BasicBlockClass.BB_IRREDUCIBLE
                self._non_back_preds.at(w).add(ydash.dfs_number)
            else:
                if ydash.dfs_number != w:
                    if not node_pool.has_some(lambda e: e == ydash):
                        work_list.append(ydash)
                        node_pool.append(ydash)

        self._non_back_preds.at(x.dfs_number).for_each(each)

    def _set_loop_attributes(self, w, node_pool, loop):
        self._nodes[w].loop = loop

        def each(node):
            self._header[node.dfs_number] = w
            node.union(self._nodes[w])

            if node.loop is not None:
                node.loop.set_parent(loop)
            else:
                loop.add_node(node.bb)

        node_pool.for_each(each)

    def _step_d(self, w, node_pool):
        def each(v):
            if v != w:
                node_pool.append(self._nodes[v].find_set())
            else:
                self._type[w] = _BasicBlockClass.BB_SELF

        self._back_preds.at(w).for_each(each)
